### PR TITLE
EdgeIndexModel and VertexIndexModel do element.load() can be parallel in performScan

### PR DESCRIPTION
**Pull request by 19kka**
io.hgraphdb.models.EdgeIndexModel#performEdgesScan parallel when element.load()

please help me review it

### CODE REVIEW

**Comment by 19kka** [comment]
@rayokota please help me review it

**Comment by 19kka** [comment]
I was changed the implementation so that it is safer when parallel

File: src/main/java/io/hgraphdb/models/EdgeIndexModel.java [hunk 1]
```diff
@@ -163,37 +165,31 @@ public void deleteEdgeIndex(Edge edge, String key, Long ts) {
...
     private Iterator<Edge> performEdgesScan(HBaseVertex vertex, Scan scan, Tuple cacheKey,
                                             boolean useIndex, Predicate<HBaseEdge> filter) {
-        List<Edge> cached = new ArrayList<>();
         final EdgeIndexReader parser = new EdgeIndexReader(graph);
         ResultScanner scanner;
         try {
             scanner = table.getScanner(scan);
-            Iterator<Edge> iterator = CloseableIteratorUtils.flatMap(
-                    CloseableIteratorUtils.concat(scanner.iterator(), IteratorUtils.of(Result.EMPTY_RESULT)),
+            List<Edge> edges = Lists.newArrayList(scanner.iterator())
```
Comment (rayokota) [inline src/main/java/io/hgraphdb/models/EdgeIndexModel.java]: Please check if `graph.isParallelLoading()` and then retain the old code if `false`.

File: src/main/java/io/hgraphdb/models/VertexIndexModel.java [hunk 2]
```diff
@@ -93,30 +97,26 @@ public void deleteVertexIndex(Vertex vertex, String key, Long ts) {
         ResultScanner scanner;
         try {
             scanner = table.getScanner(scan);
-            Iterator<Vertex> iterator = CloseableIteratorUtils.flatMap(
-                    CloseableIteratorUtils.concat(scanner.iterator(), IteratorUtils.of(Result.EMPTY_RESULT)),
+            List<Vertex> vertices = Lists.newArrayList(scanner.iterator())
```
Comment (rayokota) [inline src/main/java/io/hgraphdb/models/VertexIndexModel.java]: Please check if `graph.isParallelLoading()` and then retain the old code if `false`.

File: src/main/java/io/hgraphdb/models/EdgeIndexModel.java [hunk 3]
```diff
@@ -163,37 +165,31 @@ public void deleteEdgeIndex(Edge edge, String key, Long ts) {
...
+                                return Stream.empty();
                             }
                         } catch (final HBaseGraphNotFoundException e) {
                             edge.removeStaleIndex();
-                            return Collections.emptyIterator();
+                            return Stream.empty();
                         }
-                    });
-            return new DefaultCloseableIterator<Edge>(iterator) {
+                    }).collect(Collectors.toList());
```
Comment (rayokota) [inline src/main/java/io/hgraphdb/models/EdgeIndexModel.java]: Please add `scanner.close()`

File: src/main/java/io/hgraphdb/models/VertexIndexModel.java [hunk 4]
```diff
@@ -93,30 +97,26 @@ public void deleteVertexIndex(Vertex vertex, String key, Long ts) {
...
+                                return Stream.empty();
                             }
                         } catch (final HBaseGraphNotFoundException e) {
                             vertex.removeStaleIndex();
-                            return Collections.emptyIterator();
+                            return Stream.empty();
                         }
-                    });
-            return new DefaultCloseableIterator<Vertex>(iterator) {
+                    }).collect(Collectors.toList());
```
Comment (rayokota) [inline src/main/java/io/hgraphdb/models/VertexIndexModel.java]: Please add `scanner.close()`

File: src/main/java/io/hgraphdb/models/EdgeIndexModel.java (hunk 3 above)
Comment (19kka) [reply src/main/java/io/hgraphdb/models/EdgeIndexModel.java]: added  `scanner.close()`

File: src/main/java/io/hgraphdb/models/VertexIndexModel.java (hunk 4 above)
Comment (19kka) [reply src/main/java/io/hgraphdb/models/VertexIndexModel.java]: added  `scanner.close()`

**Comment by 19kka** [comment]
I tested the code and can execute it correctly according to the configuration `graph.isParallelLoading()`. However, when the data is very large (such as a million sides), `collect(Collectors.toList())` may cause OOM

**Comment by rayokota** [comment]
The old code would not try to load all the objects in memory. Please add an `if graph.isParallelLoading` block where the `else` block retains the old code and the `then` block is your new code.  That way users can turn off parallelization and avoid OOM.

**Comment by 19kka** [comment]
add if else block distinguish between `isParallelLoading` or not

**Pull request merged**